bundle/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/Executor.scala

```diff
@@ -0,0 +1,49 @@
+package edu.gemini.seqexec.server
+
+import edu.gemini.seqexec.server.SeqexecFailure._
+import edu.gemini.spModel.config2.{Config, ConfigSequence, ItemKey}
+import edu.gemini.spModel.obscomp.InstConstants.INSTRUMENT_NAME_PROP
+import edu.gemini.spModel.seqcomp.SeqConfigNames.INSTRUMENT_KEY
+
+import scalaz._
+import Scalaz._
+
+import scalaz.concurrent.Task
+
+/**
+ * Created by jluhrs on 4/28/15.
+ */
+final class Executor(sequenceConfig: ConfigSequence) {
+
+  def run: \/[Throwable, List[StepResult]] = runSequence(sequenceConfig.getAllSteps.toList).attemptRun
+
+  private def runSequence(config: List[Config]): Task[List[StepResult]] = {
+    (config map (runStep(_))).sequenceU
+  }
+
+  private def runStep(stepConfig: Config): Task[StepResult] = {
+    val instName = stepConfig.getItemValue(new ItemKey(INSTRUMENT_KEY, INSTRUMENT_NAME_PROP))
+    val instrument = instName match {
+      case GMOS_S.name => Some(GMOS_S)
+      case _           => None
+    }
+
+    instrument map (a => {
+      val systems: List[System] = List(TCS, a)
+      for {
+        r <- Task.gatherUnordered( systems map (x => x.configure(stepConfig)) , true)
+        s <- a.observe(stepConfig)
+      } yield StepResult(r, s)
+    }) getOrElse Task.fail(UnrecognizedInstrument(instName.toString))
```

Comment:
I think this should work fine but the issue is that every failure case has be an exception. That may be okay for this application I don't know.  I found it better to be able to enumerate the possible application failure cases (say under the trait `SeqFailure`) and add one failure case that wraps connection exceptions, etc. to turn them into an application failure case (say `case class SeqException(t: Throwable) extends SeqFailure`).

You get the fail-fast behavior with `EitherT` as in

````scala
type SeqAction[+A] = EitherT[Task, SeqFailure, A]
````

which is a combination `Task` and disjunction `\/`.  You `run` the action to get the `Task[SeqFailure \/ A]` which can then be executed with `attemptRun` or whatever to produce a result of type `\/[Throwable, \/[SeqFailure, A]]` which can be merged into a `SeqFailure \/ A` with something like:

````scala
result.fold(t => SeqException(t), identity)
````

to end up with just `SeqFailure \/ A`.

Maybe it is overkill for your application though.  Do you anticipate application level failure cases?